### ISSUE TITLE
Define Meteor methods in Class

### ISSUE THREAD

**Issue by Faberle** [jagi/meteor-astronomy#470]
What about allowing to add methods to a Class that are automatically wrapped as a Meteor Method:

``` javascript
const Post = Class.create({
  name: 'Post',
  /* ... */
  methods: {
    rename(title) {
      // Check if a given user can rename this post.
      if (this.ownerId !== Meteor.userId()) {
        throw new Meteor.Error(403, 'You are not an owner');
      }
      this.title = this;
      this.save();
     }
  }
});

Meteor.methods({
  renamePost(postId, title) {
    const post = Post.findOne(postId);
    post.rename(title);
  }});
```

becomes: 

``` javascript
Post.extend({
meteorMethods: {
  renamePost(title) {
    // Check if a given user can rename this post.
      if (this.ownerId !== Meteor.userId()) {
        throw new Meteor.Error(403, 'You are not an owner');
      }
      this.title = this;
      this.save()
  }});
```

Where this code is automatically wrapped in a Meteor method that fetches the relevant document using the _id and sets the function context accordingly.

**Comment by lukejagodzinski** [comment]
I was thinking about this feature as it's very common to just execute single function in a meteor method. So I will probably add this soon.

**Comment by jhuenges** [comment]
But what would be the benefit? Does it matter if you do ``Meteor.call({...})` or `doc.<method>()`?

**Comment by lukejagodzinski** [comment]
@jhuenges less code to write. When you define Astronomy method, appropriate Meteor wrapper method gets created so you don't have to define it by yourself. It's not about calling method, it's about defining it.

**Comment by Faberle** [comment]
What do you think about creating a naming convention for Meteor methods on Class objects, so you now if that call triggers a meteor method or not?

**Comment by jsamr** [comment]
@Faberle You can use [@server Serrurier decorator](https://github.com/sveinburne/serrurier#server-decorator) meanwhile

**Comment by arggh** [comment]
I think this is the missing piece of Astronomy.

I feel hesitant to allow updates from the client (maybe for no reason?), but find it annoying having to define these meteor methods all over the place, like loop I have to jump through.

**Comment by lukejagodzinski** [comment]
This feature will appear in the next Astronomy version

**Comment by arggh** [comment]
@jagi Any idea when that would be? I just poured myself a large cup of coffee and was about to start refactoring my code base from Astronomy V1 => V2 :)

**Comment by lukejagodzinski** [comment]
I can't tell right now because I have some extra work to do. I hope it won't take to much time.

**Comment by lukejagodzinski** [comment]
@Faberle lets just move discussion from PR to this thread.

Feature is almost ready. Big thanks for @Faberle for great help with implementation of this module. I want to discuss how it works right now and what can still be improved/changed.

**1) Ok, so at first, we can define meteor methods in the following way:**

``` js
const User = Class.create({
  name: 'User',
  collection: Users,
  fields: {
    firstName: String,
    lastName: String
  },
  meteorMethods: {
    changeName(firstName, lastName, invocation) {
      check(firstName, String);
      check(lastName, String);
      this.firstName = firstName;
      this.lastName = lastName;
      return this.save();
    }
  }
});

// Invocation
const user = User.findOne();
user.changeName('John', 'Smith', function(err, result) {
});
```

As you can see there is a new section in class definition called `meteorMethods`. We define this methods in the same way as regular methods. The only difference is that the last argument of method is `meteor method invocation object` which is just the `this` context of regular Meteor method. So we can access such properties like `this.isSimulation`, this.unblock()`or`this.userId`. We still have to directly save document if it's needed.

On method execution, we can pass callback function as the last argument.

**2) Server only method**

There may be situations when we want to hide some logic and put in server only method. Because of that it's not possible to invoke method from the client without messy workarounds like this:

``` js
// both.js
const User = Class.create({
  name: 'User',
  collection: Users,
  fields: {
    firstName: String,
    lastName: String
  }
});
// client.js
User.extend({
  meteorMethods: {
    changeName() {}
  }
})
// server.js
User.extend({
  meteorMethods: {
    changeName(firstName, lastName, invocation) {
      check(firstName, String);
      check(lastName, String);
      this.firstName = firstName;
      this.lastName = lastName;
      return this.save();
    }
  }
});
```

As you can see, we had to create dummy function on the client to have "access point" for calling method. If someone wants to use this approach, he/she is free to go. However I've also introduced two helper methods `callMethod` and `applyMethod`. Using them, we don't have to create dummy client side method. Let's take a look at the example:

``` js
// Invocation
const user = User.findOne();
user.callMethod('changeName', 'John', 'Smith', function(err, result) {
});
// or
user.applyMethod('changeName', ['John', 'Smith'], function(err, result) {
});
```

**3) What do you think?**

I'm a little bit worried about naming. Meteor methods and regular methods, it might be confusing. I was thinking that maybe regular methods should be actually Meteor methods from the beginning of Astronomy. And regular methods in the current form should be renamed to `helpers` or something similar. I'm considering introducing some option in Astronomy that would allow this naming change without breaking backward compatibility. This option would be per class, to make switching to new convention easier (class by class migration).

``` js
const User = Class.create({
  name: 'User',
  collection: Users,
  fields: {
    firstName: String,
    lastName: String
  },
  options: {
    newMethods: true
  },
  helpers: {
    fullName() {
      return `${this.firstName} ${this.lastName}`;
    }
  },
  methods: {
    changeName(firstName, lastName, invocation) {
      check(firstName, String);
      check(lastName, String);
      this.firstName = firstName;
      this.lastName = lastName;
      return this.save();
    }
  }
});
```

And also to force/encourage developers to switch to new convention I could log deprecation warning in the console.

**Comment by jhuenges** [comment]
I like the renaming suggestion. In my current project I use methods more like helpers (`getFullname()` etc.). They feel more like helpers than actual methods.

**Comment by Faberle** [comment]
I also like the renaming suggestion! And good point with the sever only methods.

**Comment by lukejagodzinski** [comment]
Ok, glad that you like this change. I've already prepared a fix that will suggest using `helpers` instead `methods`. However, for backward compatibility `methods` will become `helpers` and Meteor methods will be defined under `meteorMethods` and NOT `methods`. It can be changed in Astronomy 3.0.

**Comment by Genroa** [comment]
Thank you for this great evolution! :) server only methods will save time. I like the renaming suggestion : `helpers` for class (OOP) methods, `methods` for server only callable (Meteor) methods for this class.

**Comment by lukejagodzinski** [comment]
Feature implemented in 2.2.0. Closing this issue.
